bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/async/v2/TestAsyncExecutor.java

```diff
@@ -0,0 +1,213 @@
+    underTest.readRowsAsync(ReadRowsRequest.getDefaultInstance());
+    Assert.assertTrue(underTest.hasInflightRequests());
+    completeCall();
+    Assert.assertFalse(underTest.hasInflightRequests());
+  }
+
+  @Test
+  public void testInvalidMutation() throws Exception {
+    try {
+      when(client.mutateRowAsync(any(MutateRowRequest.class))).thenThrow(new RuntimeException());
+      underTest.mutateRowAsync(MutateRowRequest.getDefaultInstance());
+    } catch(Exception ignored) {
+    }
+    completeCall();
+    Assert.assertFalse(underTest.hasInflightRequests());
+  }
+
+  @Test
+  /**
+   * Tests to make sure that mutateRowAsync will perform a wait() if there is a bigger count of RPCs
+   * than the maximum of the RpcThrottler.
+   */
+  public void testRegisterWaitsAfterCountLimit() throws Exception {
+    ExecutorService testExecutor = Executors.newCachedThreadPool();
+    try {
+      when(client.mutateRowAsync(any(MutateRowRequest.class))).thenReturn(future);
+      // Fill up the Queue
+      for (int i = 0; i < 10; i++) {
+        underTest.mutateRowAsync(MutateRowRequest.getDefaultInstance());
+      }
+      final AtomicBoolean eleventRpcInvoked = new AtomicBoolean(false);
+      testExecutor.submit(new Callable<Void>() {
+        @Override
+        public Void call() throws Exception {
+          underTest.mutateRowAsync(MutateRowRequest.getDefaultInstance());
+          eleventRpcInvoked.set(true);
+          return null;
+        }
+      });
+      Thread.sleep(50);
+      Assert.assertFalse(eleventRpcInvoked.get());
+      completeCall();
+      Thread.sleep(50);
+      Assert.assertTrue(eleventRpcInvoked.get());
+    } finally {
+      testExecutor.shutdownNow();
+    }
+  }
+
+  @Test
+  /**
+   * Tests to make sure that mutateRowAsync will perform a wait() if there is a bigger accumulated
+   * serialized size of RPCs than the maximum of the RpcThrottler.
+   */
+  public void testRegisterWaitsAfterSizeLimit() throws Exception {
+    ExecutorService testExecutor = Executors.newCachedThreadPool();
+    try {
+      when(client.mutateRowAsync(any(MutateRowRequest.class))).thenReturn(future);
+      // Send a huge request to block further RPCs.
+      underTest.mutateRowAsync(MutateRowRequest.newBuilder()
+          .setRowKey(ByteString.copyFrom(new byte[1000])).build());
+      final AtomicBoolean newRpcInvoked = new AtomicBoolean(false);
+      Future<Void> future = testExecutor.submit(new Callable<Void>() {
+        @Override
+        public Void call() throws Exception {
+          underTest.mutateRowAsync(MutateRowRequest.getDefaultInstance());
+          newRpcInvoked.set(true);
+          return null;
+        }
+      });
+      try {
+        future.get(50, TimeUnit.MILLISECONDS);
+        Assert.fail("The future.get() call should timeout.");
+      } catch(TimeoutException expected) {
+        // Expected Exception.
+      }
+      completeCall();
+      future.get(50, TimeUnit.MILLISECONDS);
```

Comment:
We potentially need to evaluate a better alternative overall.  We spoke offline about this specific issue, and determined that CountdownLatch doesn't make sense in this scenario.